app/code/community/FrontCommerce/Payline/Helper/Duplication.php

```diff
@@ -102,11 +106,187 @@ public function getDataToken($quote)
+          }
+
+          // create invoice if needed
+          Mage::helper('payline')->automateCreateInvoiceAtShopReturn('CPT', $this->order);
+        } else {
+          // payment NOT OK
+          $msgLog = 'PAYMENT KO : ' . $webPaymentDetails['result']['code'] . ' ' . $webPaymentDetails['result']['shortMessage'] . ' (' . $webPaymentDetails['result']['longMessage'] . ')';
+          $tokenForUpdate->setResultCode($webPaymentDetails['result']['code']);
+
+          $pendingCodes = array(
+            '02306', // Customer has to fill his payment data
+            '02533', // Customer not redirected to payment page AND session is active
+            '02000', // transaction in progress
+            '02005', // transaction in progress
+            '02015', // transaction delegated to partner
+            '02016', // Transaction pending by the partner
+            '02017', // Transaction pending
+            paylineSDK::ERR_CODE // communication issue between Payline and the store
+          );
+
+          if (!in_array($webPaymentDetails['result']['code'], $pendingCodes)) {
+            if ($webPaymentDetails['result']['code'] == '00000') { // payment is OK, but a mismatch with order was detected
+              $paymentDataMismatchStatus = Mage::getStoreConfig('payment/payline_common/payment_mismatch_order_status');
+              $this->order->setState(Mage_Sales_Model_Order::STATE_CANCELED, $paymentDataMismatchStatus, $msgLog, false);
+            } elseif (in_array($webPaymentDetails['result']['code'], array('02304', '02324', '02534'))) {
+              $abandonedStatus = Mage::getStoreConfig('payment/payline_common/resignation_order_status');
+              $this->order->setState(Mage_Sales_Model_Order::STATE_CANCELED, $abandonedStatus, $msgLog, false);
+            } elseif ($webPaymentDetails['result']['code'] == '02319') {
+              $userMessage = Mage::helper('payline')->__('Your payment is canceled');
+              $canceledStatus = Mage::getStoreConfig('payment/payline_common/canceled_order_status');
+              $this->order->setState(Mage_Sales_Model_Order::STATE_CANCELED, $canceledStatus, $msgLog, false);
+            } else {
+              $userMessage = Mage::helper('payline')->__('Your payment is refused');
+              $failedOrderStatus = Mage::getStoreConfig('payment/payline_common/failed_order_status');
+              $this->order->setState(Mage_Sales_Model_Order::STATE_CANCELED, $failedOrderStatus, $msgLog, false);
+            }
+            $tokenForUpdate->setStatus(2); // KO
+            $paymentStatus = static::STATUS_ERROR;
+          } else {
+            $userMessage = Mage::helper('payline')->__('Your payment is saved');
+            $waitOrderStatus = Mage::getStoreConfig('payment/payline_common/wait_order_status');
+            $this->order->setState(Mage_Sales_Model_Order::STATE_PROCESSING, $waitOrderStatus, $msgLog, false);
+            $tokenForUpdate->setStatus(3); // to be determined
```

Comment:
put those 1, 2, 3 in a constant with a name (improves readability and avoids
those comment)